### PR TITLE
promotions-mapper

### PR DESCRIPTION
**Pull request by tatitati**
once that promotions-repository is ready then the mapper is necessary (THIS PR)

### CODE REVIEW

File: src/Mapper/ProgrammesDbToDomain/MapperFactory.php [hunk 1]
```diff
@@ -114,6 +114,15 @@ public function getMasterBrandMapper(): MasterBrandMapper
         return $this->instances['MasterBrandMapper'];
     }
 
+    public function getPromotionMapper(): PromotionMapper
```
Comment (BPScott) [inline src/Mapper/ProgrammesDbToDomain/MapperFactory.php]: Alphabetical order please.

File: src/Mapper/ProgrammesDbToDomain/PromotionMapper.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                ->getDomainModel($dbPromotion['promotionOfCoreEntity']);
+        }
+
+        if (!empty($dbPromotion['promotionOfImage'])) {
+            return $this->mapperFactory
+                ->getImageMapper()
+                ->getDomainModel($dbPromotion['promotionOfImage']);
+        }
+
+        throw new InvalidArgumentException('A promotion has to promote something, but this one is promoting nothing');
```
Comment (BPScott) [inline src/Mapper/ProgrammesDbToDomain/PromotionMapper.php]: There is a difference between `$dbPromotion['blah']` being absent, and being null. Absent means "we didn't join to that table", while set but null means "we did join to that table but and there was nothing there".  I think we are incapable of judging if we have a valid Promotion if either of those keys do not exist, so you should check that and throw a `DataNotFetchedException`. See MasterBrandMapper->getNetwork() for an example of telling the difference. Currently you're assuming both of these keys exist.

You'll want two tests to confirm what happens when you don't pass in a valid coreEntity/image and when you don't join to either of those two tables

File: tests/Mapper/ProgrammesDbToDomain/PromotionMapperTest.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            'pid' => 'p000000h',
+            'startDate' => new DateTimeImmutable('1900-01-01 00:00:00.000000'),
+            'endDate' => new DateTimeImmutable('3000-01-01 00:00:00.000000'),
+            'weighting' => 73,
+            'isActive' => 1,
+            'promotedFor' => null,
+            'title' => 'active promotion of CoreEntity',
+            'uri' => 'www.myuri.com',
+            'cascadesToDescendants' => false,
+            'createdAt' => new DateTimeImmutable('2017-08-11 10:39:25.000000'),
```
Comment (BPScott) [inline tests/Mapper/ProgrammesDbToDomain/PromotionMapperTest.php]: You can remove any values that aren't used to create the promotion object, as they're just adding noise.

File: tests/Mapper/ProgrammesDbToDomain/PromotionMapperTest.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
```
Comment (BPScott) [inline tests/Mapper/ProgrammesDbToDomain/PromotionMapperTest.php]: You can remove any values that aren't used to create the promotion object, as they're just adding noise.

File: src/Mapper/ProgrammesDbToDomain/PromotionMapper.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            $this->getSynopses($dbPromotion),
+            $dbPromotion['uri'],
+            $dbPromotion['weighting']
+        );
+    }
+
+    /**
+     * @param array[] $dbPromotion
+     */
+    private function getSynopses(array $dbPromotion): Synopses
```
Comment (votemike) [inline src/Mapper/ProgrammesDbToDomain/PromotionMapper.php]: Does this crop up all over the place? Would it be worth turning in to a trait?

Comment (tatitati) [inline src/Mapper/ProgrammesDbToDomain/PromotionMapper.php]: I thought something similar in here. But I think we can make just another PR that move all this getSynopses() methods to a trait. Otherwise my PR would be noisy I think mixing different things.
It appear (repeated in these places):

PromotionMapper
CoreEntityMapper
ImageMapper
RelatedLinkMapper
SegmentEventMapper
